--- configure    2005-12-24 04:36:33.000000000 -0800
+++ configure    2006-01-13 04:43:07.000000000 -0800
@@ -14394,7 +14394,7 @@
 	rhapsody*)	: ${LDSHARED='cc -dynamic -bundle -undefined suppress'}
 			: ${LDFLAGS=""}
 			rb_cv_dlopen=yes ;;
-	darwin*)	: ${LDSHARED='cc -dynamic -bundle -undefined suppress -flat_namespace'}
+	darwin*)	: ${LDSHARED="\$(CC) -dynamic -bundle -undefined suppress -flat_namespace"}
 			: ${LDFLAGS=""}
 			: ${LIBPATHENV=DYLD_LIBRARY_PATH}
 			rb_cv_dlopen=yes ;;
@@ -15385,6 +15385,11 @@
     arch="fat-${target_os}"

     cat >>confdefs.h <<_ACEOF
+#define __ARCHITECTURE__ "${arch}"
+_ACEOF
+
+
+    cat >>confdefs.h <<_ACEOF
 #define RUBY_THIN_ARCHLIB "${RUBY_LIB_PATH}/" __ARCHITECTURE__ "-${target_os}"
 _ACEOF
