Index: /home/toddw/src/php/xdebug/xdebug.git/php_xdebug.h
index 4ba18d0..729cc98 100644
--- a/php_xdebug.h
+++ b/php_xdebug.h
@@ -20,7 +20,7 @@
 #define PHP_XDEBUG_H
 
 #define XDEBUG_NAME       "Xdebug"
-#define XDEBUG_VERSION    "2.5.0"
+#define XDEBUG_VERSION    "2.5.0-komodo"
 #define XDEBUG_AUTHOR     "Derick Rethans"
 #define XDEBUG_COPYRIGHT  "Copyright (c) 2002-2013 by Derick Rethans"
 #define XDEBUG_COPYRIGHT_SHORT "Copyright (c) 2002-2013"
